utils/compare.ts

```diff
@@ -32,70 +33,142 @@ interface NaturalCompareOptions<T extends SortingNode>
-        formatString(nodeValueGetter(bNode)),
-        options.locales,
-      )
-  } else if (options.type === 'natural') {
-    let naturalCompare = createNaturalCompare({
-      locale: options.locales.toString(),
-    })
-    let formatString = getFormatStringFunction(
-      options.ignoreCase,
-      options.specialCharacters,
+}
+
+let getNaturalSortingFunction = <T extends SortingNode>(
+  { specialCharacters, ignoreCase, locales }: NaturalCompareOptions<T>,
+  nodeValueGetter: (node: T) => string,
+): SortingFunction<T> => {
+  let naturalCompare = createNaturalCompare({
+    locale: locales.toString(),
+  })
+  let formatString = getFormatStringFunction(ignoreCase, specialCharacters)
+  return (aNode: T, bNode: T) =>
+    naturalCompare(
+      formatString(nodeValueGetter(aNode)),
+      formatString(nodeValueGetter(bNode)),
     )
-    sortingFunction = (aNode, bNode) =>
-      naturalCompare(
-        formatString(nodeValueGetter(aNode)),
-        formatString(nodeValueGetter(bNode)),
-      )
-  } else {
-    sortingFunction = (aNode, bNode) => {
-      let aSize = aNode.size
-      let bSize = bNode.size
-
-      let { maxLineLength } = options
-
-      if (maxLineLength) {
-        let isTooLong = (size: number, node: T): undefined | boolean =>
-          size > maxLineLength && node.hasMultipleImportDeclarations
-
-        if (isTooLong(aSize, aNode)) {
-          aSize = nodeValueGetter(aNode).length + 10
-        }
-
-        if (isTooLong(bSize, bNode)) {
-          bSize = nodeValueGetter(bNode).length + 10
-        }
-      }
+}
 
-      return aSize - bSize
+let getCustomSortingFunction = <T extends SortingNode>(
+  { specialCharacters, ignoreCase, alphabet }: CustomCompareOptions<T>,
+  nodeValueGetter: (node: T) => string,
+): SortingFunction<T> => {
+  let formatString = getFormatStringFunction(ignoreCase, specialCharacters)
+  let indexByCharacters = alphabetCache.get(alphabet)
+  if (!indexByCharacters) {
+    indexByCharacters = new Map()
+    for (let [index, character] of [...alphabet].entries()) {
+      indexByCharacters.set(character, index)
     }
+    alphabetCache.set(alphabet, indexByCharacters)
+  }
+  return (aNode: T, bNode: T) => {
+    let aValue = formatString(nodeValueGetter(aNode))
+    let bValue = formatString(nodeValueGetter(bNode))
+    // Iterate character by character
+    // eslint-disable-next-line unicorn/no-for-loop
+    for (let i = 0; i < aValue.length; i++) {
+      let aCharacter = aValue[i]
+      let bCharacter = bValue[i]
+      let indexOfA = indexByCharacters.get(aCharacter)
+      let indexOfB = indexByCharacters.get(bCharacter)
+      // eslint-disable-next-line no-undefined
+      if (indexOfA === undefined) {
+        indexOfA = Infinity
+      }
+      // eslint-disable-next-line no-undefined
+      if (indexOfB === undefined) {
+        indexOfB = Infinity
+      }
+      if (indexOfA !== indexOfB) {
+        return indexOfA - indexOfB > 0 ? 1 : -1
```

Comment:
Perhaps we should create a utility in the future:

```ts
let convertBooleanToSign = (boolean: boolean): number => boolean ? 1 : -1
```

Because we have these numbers repeating in a lot of places.

Reply:
@azat-io 👍 I added the function in 19eec7b7c59855ab70176a18f12d5efa89f26667.